xonsh/proc.py

```diff
@@ -541,14 +542,36 @@ def _wcode_to_popen(code):
                                    "stdin_redirect",
                                    "stdout_redirect",
                                    "stderr_redirect",
-                                   "timestamp"])
+                                   "timestamp",
+                                   "executed_cmd"])
+
 
 class CompletedCommand(_CCTuple):
     """Represents a completed subprocess-mode command."""
 
     def __bool__(self):
         return self.returncode == 0
 
+    def __iter__(self):
+        if not self.stdout:
+            raise StopIteration()
+
+        pre = self.stdout
+        post = None
+
+        while post != '':
+            pre, sep, post = pre.partition('\n')
+            # this line may be optional since we use universal newlines.
+            pre = pre[:-1] if pre and pre[-1] == '\r' else pre
+            yield pre
+            pre = post
+
+        if self.returncode:
+            # I included self, as providing access to stderr and other details
+            # useful when instance isn't assigned to a variable in the shell.
+            raise XonshCalledProcessError(self.returncode, self.executed_cmd,
+                                          self.stdout, self.stderr, self)
```

Comment:
Also, based on some of the gitter/irc chat I think that the raising part of this should be moved to another method, something like:

```python
def itercheck(self):
    yield from self
    if self.returncode:
        rasise ...
```